src/creatures/players/cast/cast_viewer.cpp

```diff
@@ -0,0 +1,1533 @@
+	if (m_owner) {
+		return m_owner->canSee(pos);
+	}
+
+	return false;
+}
+
+uint32_t CastViewer::getIP() const {
+	if (m_owner) {
+		return m_owner->getIP();
+	}
+
+	return 0;
+}
+
+void CastViewer::sendStats() {
+	if (m_owner) {
+		m_owner->sendStats();
+
+		for (const auto &it : m_viewers) {
+			it.first->sendStats();
+		}
+	}
+}
+
+void CastViewer::sendPing() {
+	if (m_owner) {
+		m_owner->sendPing();
+
+		for (const auto &it : m_viewers) {
+			it.first->sendPing();
+		}
+	}
+}
+
+void CastViewer::logout(bool displayEffect, bool forceLogout) {
+	if (m_owner) {
+		m_owner->logout(displayEffect, forceLogout);
+	}
+}
+
+void CastViewer::sendAddContainerItem(uint8_t cid, uint16_t slot, std::shared_ptr<Item> item) {
+	if (m_owner) {
+		m_owner->sendAddContainerItem(cid, slot, item);
+
+		for (const auto &it : m_viewers) {
+			it.first->sendAddContainerItem(cid, slot, item);
+		}
+	}
+}
+
+void CastViewer::sendUpdateContainerItem(uint8_t cid, uint16_t slot, std::shared_ptr<Item> item) {
+	if (m_owner) {
+		m_owner->sendUpdateContainerItem(cid, slot, item);
+
+		for (const auto &it : m_viewers) {
+			it.first->sendUpdateContainerItem(cid, slot, item);
+		}
+	}
+}
+
+void CastViewer::sendRemoveContainerItem(uint8_t cid, uint16_t slot, std::shared_ptr<Item> lastItem) {
+	if (m_owner) {
+		m_owner->sendRemoveContainerItem(cid, slot, lastItem);
+
+		for (const auto &it : m_viewers) {
+			it.first->sendRemoveContainerItem(cid, slot, lastItem);
+		}
+	}
+}
+
+void CastViewer::sendUpdatedVIPStatus(uint32_t guid, VipStatus_t newStatus) {
+	if (m_owner) {
+		m_owner->sendUpdatedVIPStatus(guid, newStatus);
+
+		for (const auto &it : m_viewers) {
+			it.first->sendUpdatedVIPStatus(guid, newStatus);
+		}
+	}
+}
+
+void CastViewer::sendVIP(uint32_t guid, const std::string &name, const std::string &description, uint32_t icon, bool notify, VipStatus_t status) {
+	if (m_owner) {
+		m_owner->sendVIP(guid, name, description, icon, notify, status);
+
+		for (const auto &it : m_viewers) {
+			it.first->sendVIP(guid, name, description, icon, notify, status);
+		}
+	}
+}
+
+void CastViewer::sendVIPGroups() {
+	if (m_owner) {
+		m_owner->sendVIPGroups();
+
+		for (const auto &it : m_viewers) {
+			it.first->sendVIPGroups();
+		}
+	}
+}
+
+void CastViewer::sendClosePrivate(uint16_t channelId) {
+	if (m_owner) {
+		m_owner->sendClosePrivate(channelId);
+	}
+}
+
+void CastViewer::sendFYIBox(const std::string &message) {
+	if (m_owner) {
+		m_owner->sendFYIBox(message);
+	}
+}
+
+uint32_t CastViewer::getVersion() const {
+	if (m_owner) {
+		return m_owner->getVersion();
+	}
+
+	return 0;
+}
+
+void CastViewer::disconnect() {
+	if (m_owner) {
+		m_owner->sendSessionEndInformation(SESSION_END_LOGOUT);
+	}
+}
+
+void CastViewer::sendCreatureSkull(std::shared_ptr<Creature> creature) const {
+	if (m_owner) {
+		m_owner->sendCreatureSkull(creature);
+
+		for (const auto &it : m_viewers) {
+			it.first->sendCreatureSkull(creature);
+		}
+	}
+}
```

Comment:
Hello @jhogberg, thank you very much for your detailed and insightful feedback!

You are absolutely right in pointing out that frequently passing std::shared_ptr by value, especially in functions that do not require ownership of the object, is not ideal due to the costs associated with atomic operations on the reference counter. Your suggestion to pass std::shared_ptr by const reference in cases where the function does not take ownership of the object is an excellent practice that we will definitely adopt.

I will review parts of the code where this practice can be applied to enhance performance and efficiency. Thanks again for bringing this to our attention!

Reply:
https://github.com/opentibiabr/canary/pull/2863/files

Reply:
Great :)

If I'm reading the code right there's another measure you can take for better performance, which is a lot more work but I think your current architecture could be adapted to it: [epoch-based reclamation](https://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-579.pdf) (section 5.2.3). If you're the practical sort you might prefer to read https://github.com/erlang/otp/blob/master/erts/emulator/internal_doc/ThreadProgress.md and the accompanying implementation in https://github.com/erlang/otp/blob/master/erts/emulator/beam/erl_thr_progress.c

It gives you multi-threaded deletion safety without having to deal with reference counting, "free" memory barriers, and also all the nice parts of having a garbage collector with regards to lock-free algorithms (e.g. the [ABA problem](https://en.wikipedia.org/wiki/ABA_problem) becomes moot in scenarios like the compare-exchange loops in the push/pop operations in #2825) but without the issues that come with a GC.

Since most operations seem to go through `Dispatcher` I think that's a natural place to drive this stuff. It does require a change in how you think about most things, but opens the door for much better multi-core performance.